### PR TITLE
BH-1103: Number field validation reports inaccurate errors

### PR DESCRIPTION
**Pull request by matthewguywright**
## Description

https://wpengine.atlassian.net/browse/

### CODE REVIEW

**Comment by mindctrl** [comment]
Linking to comment on the previous PR: https://github.com/wpengine/atlas-content-modeler/pull/199#issuecomment-884882727

**Review by ChrisWiegman — DISMISSED** [review]

**Comment by anthonyburchell** [comment]
I noticed that you are able to use `.0` in integer fields. 

to reproduce, create a new field set to integer and enter `5.0` for the value. This is the same as `5` technically, but I wonder if the decimal will trip folks up on the front end.

**Comment by anthonyburchell** [comment]
Could we potentially do an `onChange` to just not allow decimals to be entered in integer number fields?

**Comment by mindctrl** [comment]
@matthewguywright nice fix on the decimals getting cut off on the error message.

Something @apmatthews said yesterday is worth considering... https://github.com/wpengine/atlas-content-modeler/pull/199#discussion_r674339038

Let me show an example:
Number field, configured as Integer, advanced settings as follows: min: -5, max: 25, step: 5

On publisher side, if someone enters "2" in this field, the error message says "Step value is 5", which is confusing and helpful to understand what type of values can be entered.

Screenshot
image

Thoughts on how to address this in a way that's easy for users to understand? We can do this in another PR if we want more time to think it through.

**Comment by mindctrl** [comment]
@matthewguywright this seems to be working well now. Can you add some e2e tests that confirm the desired behavior?

File: includes/publisher/js/src/components/Field.jsx [hunk 1]
```diff
@@ -43,6 +43,8 @@ export default function Field(props) {
 					),
 					event.target.step.toString()
 				);
+			} else if (event.target.validity.customError) {
+				error = "The input is invalid.";
```
Comment (ChrisWiegman) [inline includes/publisher/js/src/components/Field.jsx:47]: This needs to be translatable.

**Review by ChrisWiegman — CHANGES_REQUESTED** [review]
This seems to work for me. Only issue is a string that is missing i18n.

File: includes/publisher/js/src/components/Field.jsx (hunk 1 above)
Comment (matthewguywright) [reply includes/publisher/js/src/components/Field.jsx]: Updated

**Review by ChrisWiegman — DISMISSED** [review]

File: tests/acceptance/PublishModelCest.php [hunk 2]
```diff
@@ -2,15 +2,86 @@
...
+		$i->wait(2);
+
+		$i->see('This field is required');
+		$i->wait(1);
+
+		$i->seeInField('atlas-content-modeler[goose][integer]', '');
+		$i->seeInField('atlas-content-modeler[goose][decimal]', '');
+	}
+
+	public function error_is_triggered_for_invalid_value(\AcceptanceTester $i)
```
Comment (mindctrl) [inline tests/acceptance/PublishModelCest.php:49]: @matthewguywright let's rename these test methods to something more descriptive and related to the number field things they're testing. Example: `i_see_submission_errors_in_number_fields_when_input_is_invalid_for_the_number_type` or something along those lines.

File: includes/publisher/js/src/components/Field.jsx [hunk 3]
```diff
@@ -24,6 +24,30 @@ export default function Field(props) {
...
+			} else if (event.target.validity.stepMismatch) {
+				error = sprintf(
+					__(
+						"Value must be a multiple of %s.",
+						"atlas-content-modeler"
+					),
+					event.target.step.toString()
+				);
+			} else if (event.target.validity.customError) {
+				error = __("The input is invalid.", "atlas-content-modeler");
```
Comment (mindctrl) [inline includes/publisher/js/src/components/Field.jsx:47]: Can we make this error message clearer? Can we tell the user why it's invalid?

**Review by mindctrl — CHANGES_REQUESTED** [review]

File: tests/acceptance/PublishModelCest.php [hunk 4]
```diff
@@ -2,15 +2,52 @@
...
 {
-	public function i_can_publish_a_model_entry(AcceptanceTester $i)
+	public function _before(\AcceptanceTester $i)
 	{
 		$i->maximizeWindow();
 
 		// First we create a model with fields.
 		$i->loginAsAdmin();
-		$i->haveContentModel('goose', 'geese', 'Geese go honk');
+		$i->haveContentModel('moose', 'mooses', 'Mooses go frmh');
```
Comment (mindctrl) [inline tests/acceptance/PublishModelCest.php:11]: It'd be nice to not change these tests from goose to moose. I know that came about due to local test failures, where the content models weren't being wiped out between tests, but I think we are addressing that another way.

Comment (matthewguywright) [reply tests/acceptance/PublishModelCest.php]: Updated

**Comment by mindctrl** [comment]
@matthewguywright if you bring this up to date with `main` the failing unit tests should pass.

**Review by anthonyburchell — DISMISSED** [review]
Awesome stuff! Great workarounds to the edge cases! LGTM 👍

File: tests/acceptance/PublishModelCest.php [hunk 5]
```diff
@@ -2,15 +2,52 @@
...
 {
-	public function i_can_publish_a_model_entry(AcceptanceTester $i)
+	public function _before(\AcceptanceTester $i)
 	{
 		$i->maximizeWindow();
 
 		// First we create a model with fields.
 		$i->loginAsAdmin();
-		$i->haveContentModel('goose', 'geese', 'Geese go honk');
+		$i->haveContentModel('goose', 'gooses', 'gooses go frmh');
```
Comment (mindctrl) [inline tests/acceptance/PublishModelCest.php:11]:
```suggestion
		$i->haveContentModel('goose', 'geese', 'Geese go honk');
```

...to get it back to what it was before... gooses? :laughing:

File: tests/acceptance/PublishModelCest.php [hunk 6]
```diff
@@ -2,15 +2,52 @@
...
+		$i->wait(1);
+
+		$i->fillField(['name' => 'atlas-content-modeler[goose][integer]'], '');
+		$i->fillField(['name' => 'atlas-content-modeler[goose][decimal]'], '');
+		$i->scrollTo('#submitdiv');
+
+		$i->click('Publish', '#publishing-action');
+		$i->wait(2);
+
+		$i->see('This field is required');
```
Comment (mindctrl) [inline tests/acceptance/PublishModelCest.php:42]: @matthewguywright I wonder if we should test a scenario where the field is not required? The Jira issue described a case where the "This field is required" message showed when it should not. A test that confirms this would catch any regressions for that specific bug.

Comment (matthewguywright) [reply tests/acceptance/PublishModelCest.php:42]: Haha. Updated.

**Review by anthonyburchell — APPROVED** [review]
Nice callout on that test case @mindctrl ! Looks good to me! :D

**Review by mindctrl — APPROVED** [review]
